lmp/generate-static-deltas

```diff
@@ -39,19 +40,19 @@ class ProgressCb:
             self.next_percent += 5
 
 
-def drain(progress: Progress, u, prog_cb: ProgressCb, dst: str, tok_secret_name: str):
+def drain(progress: Progress, u, prog_cb: ProgressCb.cb, dst: str, tok_secret_name: str):
     r = http_get(u, headers={
         "OSF-TOKEN": secret(tok_secret_name),
         "Connection": "keep-alive",
         "Keep-Alive": "timeout=1200, max=1"  # keep connection alive for 1 request for 20m
     }, stream=True)
-    p = Popen(["tar", "-xj"], cwd=dst, stdin=PIPE)
-    for chunk in r.iter_content(chunk_size=1024 * 1024):
-        if chunk:
-            p.stdin.write(chunk)
-            prog_cb(len(chunk))
-    p.stdin.close()
-    p.wait()
+    last_pos = 0
+    with io.BufferedReader(r.raw, buffer_size=1024 * 1024) as br:
+        with tarfile.open(fileobj=br, mode="r|bz2") as tar_stream:
+            for member in tar_stream:
+                tar_stream.extract(member, dst)
+                prog_cb(br.tell() - last_pos)
+                last_pos = br.tell()
```

Comment:
i can't remember why i used Popen here. I think i was getting some streaming issues, but maybe that was from the connection stuff you fixed in the other commit.

Reply:
Maybe. I checked the given implementation for the same cases when the current one fails and it works, and works faster.